cmd/influxd/run/server.go

```diff
@@ -436,88 +465,88 @@ func (s *Server) Open() error {
 	s.SnapshotterService.WithLogger(s.Logger)
 	s.Monitor.WithLogger(s.Logger)
 
-	// Open TSDB store.
-	if err := s.TSDBStore.Open(); err != nil {
-		return fmt.Errorf("open tsdb store: %s", err)
-	}
-
-	// Open the subscriber service
-	if err := s.Subscriber.Open(); err != nil {
-		return fmt.Errorf("open subscriber: %s", err)
-	}
-
-	// Open the points writer service
-	if err := s.PointsWriter.Open(); err != nil {
-		return fmt.Errorf("open points writer: %s", err)
-	}
+	// append TSDB Store, Subcriber, and PointsWriter services to our Services slice.
+	s.Services = append(s.Services, s.TSDBStore, s.Subscriber, s.PointsWriter)
 
 	s.PointsWriter.AddWriteSubscriber(s.Subscriber.Points())
 
+	// asyncronously start all services. we run this loop in a go routine so we
+	// can block on the supplied context as soon as possible.  FIXME: this might
+	// be unnessesary.
+	sem := make(chan struct{}, len(s.Services))
 	for _, service := range s.Services {
-		if err := service.Open(); err != nil {
-			return fmt.Errorf("open service: %s", err)
-		}
+		go func(svc Service) {
+			// start service
+			//
+			// we begin by "incrementing" our semaphore to indicate that a service
+			// has started.
+			//
+			// as services exit, we read back from the same channel to remove our
+			// place in the queue.
+			//
+			// to test if all services are complete, we can attempt to write to the
+			// channel len(s.Services) times which will block until all services have
+			// exited.
+			sem <- struct{}{}
+			if err := svc.Run(ctx, reg); err != nil {
+				// if there was an error, report it to s.err s.err <- err
+				//
+				// subtle: we want to report errors to s.err but if nothing is
+				// receiving messages on that channel, we do not want to block!
+				select {
+				case s.err <- err:
+				default:
+				}
+			}
+			// indicate that this service is done by decrementing our semaphore.
+			<-sem
+		}(service)
 	}
 
 	// Start the reporting service, if not disabled.
 	if !s.reportingDisabled {
-		go s.startServerReporting()
+		go s.startServerReporting(ctx)
 	}
 
-	return nil
-}
+	<-reg.WaitFor("http")
+	<-reg.WaitFor("tsdb")
 
-// Close shuts down the meta and data stores and all services.
-func (s *Server) Close() error {
+	// signal that we're ready
+	if ready != nil {
+		close(ready)
+	}
+
+	// wait for a cancellation signal
+	<-ctx.Done()
 	s.stopProfile()
 
 	// Close the listener first to stop any new connections
 	if s.Listener != nil {
 		s.Listener.Close()
 	}
 
-	// Close services to allow any inflight requests to complete
-	// and prevent new requests from being accepted.
-	for _, service := range s.Services {
-		service.Close()
+	// wait for services to complete.
+	for i, end := 0, len(s.Services); i < end; i++ {
```

Comment:
Let's simplify this a bit, for clarity?

```go
for _ = range s.Services {
    // receive signal when service has ended.
    sem <- struct{}{}
}
```

Reply:
d'oh.  i'm not sure why i did it this way.

Reply:
changed to:

```
for range s.Services {
  ...
}
```

Reply:
ha indeed! 🤦‍♂